Setup Celery as part of Django setup

millerdev opened the pull request:
In addition to a bit of cleanup, this allows `manage.py` to import things from `corehq` without triggering Celery initialization as a side effect.

:blowfish: Review by commit.

## Safety Assurance

### Safety story

Tested on staging. Web workers and celery processes appear to be working normally.

### Automated test coverage

No new test coverage.

### QA Plan

No QA.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

millerdev committed:
Remove pickle patch

Obsolete since upgrading to Python 3.9

millerdev committed:
Setup Celery as part of Django setup

This allows manage.py to import things from corehq without triggering celery initialization logic as a side effect.

millerdev committed:
Remove unneeded AppRegistryNotReady handler

millerdev committed:
Remove hack, django.setup() is idempotent

dannyroberts commented on manage.py:
```diff
@@ -115,25 +113,13 @@ def run_patches():
-    Remove after upgrading to Python 3.8+
-    """
-    import pickle
-    if pickle.HIGHEST_PROTOCOL < 5:
```
Will we need this function again the next time a python upgrade includes a new highest pickle protocol?

millerdev replied:
If we do I know where to dig it up. I hope we are upgraded to a new version of Celery with better behavior by the time we would need it.

dannyroberts commented:
Were the tests manually cancelled, or did something happen?

millerdev committed:
Guard against default celery app

Also initialize celery earlier in the Django setup process before tasks are imported.

millerdev committed:
Patch to prevent ignored error with retry in test

The sync_user_phone_numbers task catches Exception and retries 10 times, which caused a long cascade of ignored errors:

AssertionError: Database queries to 'default' are not allowed in SimpleTestCase subclasses.

millerdev committed:
Add docs and fix comments

millerdev commented:
> Were the tests manually cancelled, or did something happen?

They were manually cancelled. The problem was caused by Celery tasks being imported before the Celery app was initialized, causing them to become associated with an automatically created "default" Celery app, which caused `corehq.apps.users.tests.test_signals:TestUserSignals.test_webuser_save` to hang in `corehq/apps/analytics/signals.py:user_save_callback` on `identify.delay(couch_user.username, properties)`. I think the fact that the default Celery app had `task_always_eager == False` was related to the hang, but I didn't dig into the precise reason once I realized that the default app was being used when it should not have been.

millerdev committed:
Reset Celery app created by gevent monkey patching

This happens for workers run with a command like

    celery -A corehq worker -P gevent
